src/control/XournalMain.cpp

```diff
@@ -560,3 +415,146 @@ void XournalMain::initResourcePath(GladeSearchpath* gladePath, const gchar* rela
+            XojMsgBox::showErrorToUser(static_cast<GtkWindow*>(*app_data->win), msg);
+        }
+
+        fs::path p(app_data->optFilename[0]);
+
+        try {
+            if (fs::exists(p)) {
+                opened = app_data->control->openFile(p, app_data->openAtPageNumber);
+            } else {
+                opened = app_data->control->newFile("", p);
+            }
+        } catch (fs::filesystem_error const& e) {
+            string msg = FS(_F("Sorry, Xournal++ cannot open remote files at the moment.\n"
+                               "You have to copy the file to a local directory.") %
+                            p.u8string().c_str() % e.what());
+            XojMsgBox::showErrorToUser(static_cast<GtkWindow*>(*app_data->win), msg);
+        }
+    }
+
+    app_data->control->getScheduler()->start();
+
+    if (!opened) {
+        app_data->control->newFile();
+    }
+
+    checkForErrorlog();
+    checkForEmergencySave(app_data->control.get());
+
+    // There is a timing issue with the layout
+    // This fixes it, see #405
+    Util::execInUiThread([=]() { app_data->control->getWindow()->getXournal()->layoutPages(); });
+    gtk_application_add_window(GTK_APPLICATION(application), GTK_WINDOW(app_data->win->getWindow()));
+}
+
+void on_shutdown(GApplication* application, XMPtr app_data) {
+    app_data->control->saveSettings();
+    app_data->win->getXournal()->clearSelection();
+    app_data->control->getScheduler()->stop();
+}
+
+gint on_handle_local_options(GApplication* app, GVariantDict* options, XMPtr app_data) {
+    if (app_data->showVersion) {
+        std::cout << PROJECT_NAME << " " << PROJECT_VERSION << std::endl;
+        std::cout << "└──libgtk: " << gtk_get_major_version() << "."  //
+                  << gtk_get_minor_version() << "."                   //
+                  << gtk_get_micro_version() << std::endl;            //
+        return 0;
+    }
+
+    if (app_data->pdfFilename && app_data->optFilename && *app_data->optFilename) {
+        return exportPdf(*app_data->optFilename, app_data->pdfFilename);
+    }
+    if (app_data->imgFilename && app_data->optFilename && *app_data->optFilename) {
+        return exportImg(*app_data->optFilename, app_data->imgFilename);
+    }
+    return -1;
+}
+
+auto XournalMain::run(int argc, char** argv) -> int {
+
+    XournalMainPrivate app_data;
+    GtkApplication* app = gtk_application_new("com.github.xournalpp.xournalpp", APP_FLAGS);
+
+    g_signal_connect(app, "activate", G_CALLBACK(&on_activate), &app_data);
+    g_signal_connect(app, "command-line", G_CALLBACK(&on_command_line), &app_data);
+    g_signal_connect(app, "open", G_CALLBACK(&on_open_files), &app_data);
+    g_signal_connect(app, "startup", G_CALLBACK(&on_startup), &app_data);
+    g_signal_connect(app, "shutdown", G_CALLBACK(&on_shutdown), &app_data);
+    g_signal_connect(app, "handle-local-options", G_CALLBACK(&on_handle_local_options), &app_data);
+
+    std::array<GOptionEntry, 6> options = {
```

Comment:
Use a plain const array, no advantage to a C++ array here. We would have to change the size here every time we add/remove flags.

Reply:
The disadvantage is, that clang-tidy complains. Marking everything as `// NOLINT(<xyz>)` is only applicable, if it can't be fixed due to a dependency. I would propose to either wait till c++20 to use `std::make_array` or implement it.

Reply:
Looks like the easiest option is to just leave it as `std::array`.

Reply:
Turns out writing `std::array options = <initializer-list>` does the job! :)